src/net/client/xfr.rs

```diff
@@ -0,0 +1,1123 @@
+//! Parsing of AXFR/IXFR response messages for higher level processing.
+//!
+//! This module provides [`XfrResponseProcessor`] which enables you to process
+//! one or more AXFR/IXFR response messages in terms of the high level
+//! [`XfrEvent`]s that they represent without having to deal with the
+//! AXFR/IXFR protocol details.
+use std::fmt::Debug;
+
+use bytes::Bytes;
+use tracing::trace;
+
+use crate::base::iana::Opcode;
+use crate::base::wire::ParseError;
+use crate::base::{
+    Message, ParsedName, Record, RecordSection, Rtype, Serial,
+};
+use crate::rdata::{Soa, ZoneRecordData};
+
+//------------ XfrRecord ------------------------------------------------------
+
+/// The type of record processed by [`XfrResponseProcessor`].
+pub type XfrRecord =
+    Record<ParsedName<Bytes>, ZoneRecordData<Bytes, ParsedName<Bytes>>>;
+
+//------------ XfrResponseProcessor -------------------------------------------
+
+/// An AXFR/IXFR response processor.
+///
+/// [`XfrResponseProcessor`] can be invoked on one ore more sequentially
+/// AXFR/IXFR received response messages to verify them and during processing
+/// emit events which an implementor of [`XfrEventHandler`] can handle.
+pub struct XfrResponseProcessor<T: XfrEventHandler> {
+    evt_handler: T,
+
+    mode: Mode,
+}
+
+impl<T: XfrEventHandler> XfrResponseProcessor<T> {
+    /// Create a new XFR response processor.
+    ///
+    /// Events will be emitted to the given [`XfrEventHandler`] implementation.
+    pub fn new(evt_handler: T) -> Self {
+        Self {
+            evt_handler,
+            mode: Mode::default(),
+        }
+    }
+
+    /// Process a single AXFR/IXFR response message.
+    ///
+    /// During processing events will be emitted to the registered
+    /// [`XfrEventHandler`] for handling.
+    ///
+    /// Returns Ok(true) if the XFR response was the last in the seqence,
+    /// Ok(false) if more XFR response messages are needed to complete the
+    /// sequence, or Err on error.
+    pub async fn answer(
+        &mut self,
+        req: &Message<Bytes>,
+        resp: Message<Bytes>,
+    ) -> Result<bool, Error> {
+        // Check that the given message is a DNS XFR response.
+        let res = self.check_is_xfr_answer(req, &resp).await;
+
+        // Unpack the XFR type and answer object. We cannot do this in the
+        // line above using `map_err()` and `?` as the Rust compiler complains
+        // about attempting to return `resp` while a reference to it still
+        // exists.
+        let (xfr_type, answer) = match res {
+            Ok(values) => values,
+            Err(err) => return Err(Error::from_check_error(resp, err)),
+        };
+
+        // Walk over all records that are of the type that can exist in a zone
+        // file. TODO: Should this actually walk over all possible record
```

Comment:
In response to the TODO question: It probably should indeed process everything – it’s more important that the transferred zone is the same as the original than that it is correct. One option is to use `AllRecordData`, another option (probably better) is to make the type generic over a `ParseAnyRecordData<..>` type.

Reply:
I briefly tried `AllRecordData` but couldn't work out how to use it in this context. How would the `ParseAnyRecordData<..>` idea look, if you can give a bit more explanation of what you have in mind that would be appreciated.

Reply:
I think you only need to swap out the use of `RecordSection::limit_to` with `RecordSection::into_records` and everything else should stay the same. Might be missing something, though.

Reply:
I'm re-opening this conversation because I have for now changed it back to ``ZoneRecordData` because that's all that the zone tree supports. I didn't want to expand the scope to include zone tree changes at this point, if we decide that `AllRecordData` is what we really want. I have not yet looked at how `ZoneRecordData` and `AllRecordData` actually differ so am not able to make a decision on this yet.

Reply:
I'm going to resolve this conversation as in the time available I'm not going to adjust the underlying zone tree to handle more than `ZoneRecordData` and this PR is intended to be used to help feed data into that zone tree, i.e. has to be compatible with it.